db/rocksdb/iterator.go

```diff
@@ -0,0 +1,149 @@
+package rocksdb
+
+import (
+	"bytes"
+
+	dbm "github.com/cosmos/cosmos-sdk/db"
+	"github.com/tecbot/gorocksdb"
+)
+
+type rocksDBIterator struct {
+	source     *gorocksdb.Iterator
+	start, end []byte
+	isReverse  bool
+	isInvalid  bool
+	// Whether iterator has been advanced to the first element (is fully initialized)
+	primed bool
+}
+
+var _ dbm.Iterator = (*rocksDBIterator)(nil)
+
+func newRocksDBIterator(source *gorocksdb.Iterator, start, end []byte, isReverse bool) *rocksDBIterator {
+	if isReverse {
+		if end == nil {
+			source.SeekToLast()
+		} else {
+			source.Seek(end)
+			if source.Valid() {
+				eoakey := moveSliceToBytes(source.Key()) // end or after key
+				if bytes.Compare(end, eoakey) <= 0 {
+					source.Prev()
+				}
+			} else {
+				source.SeekToLast()
+			}
+		}
+	} else {
+		if start == nil {
+			source.SeekToFirst()
+		} else {
+			source.Seek(start)
+		}
+	}
+	return &rocksDBIterator{
+		source:    source,
+		start:     start,
+		end:       end,
+		isReverse: isReverse,
+		isInvalid: false,
+		primed:    false,
+	}
+}
+
+// Domain implements Iterator.
+func (itr *rocksDBIterator) Domain() ([]byte, []byte) {
+	return itr.start, itr.end
+}
+
+// Valid implements Iterator.
+func (itr *rocksDBIterator) Valid() bool {
+	if !itr.primed {
+		return false
+	}
+
+	// Once invalid, forever invalid.
+	if itr.isInvalid {
+		return false
+	}
+
+	// If source is invalid, invalid.
+	if !itr.source.Valid() {
+		itr.isInvalid = true
+		return false
+	}
+
+	var (
+		start = itr.start
+		end   = itr.end
+		key   = moveSliceToBytes(itr.source.Key())
+	)
+	// If key is end or past it, invalid.
+	if itr.isReverse {
+		if start != nil && bytes.Compare(key, start) < 0 {
+			itr.isInvalid = true
+			return false
+		}
+	} else {
+		if end != nil && bytes.Compare(end, key) <= 0 {
```

Comment:
I'd swap this to keep a consistent reading order from above by making it: bytes.Compare(key, end) > 0